### PR TITLE
Added Effort plugin

### PR DESCRIPTION
**Pull request by ahcorde**
Added Effort plugin

effort_plugin

### CODE REVIEW

File: rviz_default_plugins/include/rviz_default_plugins/displays/effort/effort_display.hpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+  rviz_common::properties::Property * category_;
+  rviz_common::properties::FloatProperty * effort_property_;
+  rviz_common::properties::FloatProperty * max_effort_property_;
+};
+//
+// typedef std::set<JointInfo*> S_JointInfo;
+// typedef std::vector<std::string> V_string;
+
+// class EffortVisual;
```
Comment (clalancette) [inline rviz_default_plugins/include/rviz_default_plugins/displays/effort/effort_display.hpp]: I think we should delete these commented-out lines.

File: rviz_default_plugins/include/rviz_default_plugins/displays/effort/effort_display.hpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  // Overrides of public virtual functions from the Display class.
+  void onInitialize() override;
+  void reset() override;
+
+  void load(const rviz_common::Config & config) override;
+
+  void update(float wall_dt, float ros_dt) override;
+
+private Q_SLOTS:
+//   // Helper function to apply color and alpha to all visuals.
```
Comment (clalancette) [inline rviz_default_plugins/include/rviz_default_plugins/displays/effort/effort_display.hpp]:
```suggestion
  // Helper function to apply color and alpha to all visuals.
```

File: rviz_default_plugins/plugins_description.xml [hunk 3]
```diff
@@ -25,6 +25,17 @@
     <message_type>sensor_msgs/msg/CompressedImage</message_type>
   </class>
 
+  <class
+    name="rviz_default_plugins/Effort"
+    type="rviz_default_plugins::displays::EffortDisplay"
+    base_class_type="rviz_common::Display"
+  >
+    <description>
+      The Effort display shows a sensor_msgs/JointState/effort message as circled arrows around each revolute joint in the robot. &lt;a href="http://www.ros.org/wiki/rviz/DisplayTypes/Effort"&gt;More Information&lt;/a&gt;.
```
Comment (clalancette) [inline rviz_default_plugins/plugins_description.xml]: I know we have the links to the wiki elsewhere in this file, but I don't think we should add in more.  The wiki will (eventually) go away.

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+}
+
+JointInfo::~JointInfo()
+{
+}
+
+JointInfo * EffortDisplay::createJoint(const std::string & joint)
+{
+  JointInfo * info = new JointInfo(joint, joints_category_);
+  joints_.insert(std::make_pair(joint, info));
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]: I think you need `#include <utility>` at the top of the file for `std::make_pair`.

More importantly, what is responsible for freeing the `info` pointer here?

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+  setStatus(rviz_common::properties::StatusProperty::Ok, "URDF", "Robot model parsed Ok");
+
+  for (std::map<std::string, urdf::JointSharedPtr>::iterator it = robot_model_->joints_.begin();
+    it != robot_model_->joints_.end(); it++)
+  {
+    urdf::JointSharedPtr joint = it->second;
+    if (joint->type == urdf::Joint::REVOLUTE || joint->type == 2) {
+      std::string joint_name = it->first;
+      urdf::JointLimitsSharedPtr limit = joint->limits;
+      joints_[joint_name] = createJoint(joint_name);
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]: It seems a little weird to me that `createJoint` is inserting the object into `joints_`, but then we are doing it again here.  Do we need to do that twice?

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+{
+}
+
+void JointInfo::setEffort(double e)
+{
+  effort_property_->setFloat(e);
+  effort_ = e;
+}
+void JointInfo::setMaxEffort(double m)
+{
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]:
```suggestion
}

void JointInfo::setMaxEffort(double m)
{
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+
+EffortDisplay::EffortDisplay()
+: rviz_common::MessageFilterDisplay<sensor_msgs::msg::JointState>()
+{
+  alpha_property_ = new rviz_common::properties::FloatProperty(
+    "Alpha", 1.0, "0 is fully transparent, 1.0 is fully opaque.",
+    this, SLOT(updateColorAndAlpha()));
+
+  width_property_ = new rviz_common::properties::FloatProperty(
+    "Width", 0.02, "Width to drow effort circle", this,
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]:
```suggestion
    "Width", 0.02, "Width to draw effort circle", this,
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+    SLOT(updateColorAndAlpha()));
+
+  scale_property_ = new rviz_common::properties::FloatProperty(
+    "Scale", 1.0, "Scale to drow effort circle", this,
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]:
```suggestion
    "Scale", 1.0, "Scale to draw effort circle", this,
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+    SLOT(updateHistoryLength()));
+
+  history_length_property_->setMin(1);
+  history_length_property_->setMax(100000);
+
+  robot_description_property_ =
+    new rviz_common::properties::StringProperty(
+    "Robot Description", "/robot_description",
+    "Name of the parameter to search for to load the robot "
+    "description.",
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]:
```suggestion
    "Name of the topic from which to load the robot "
    "description.",
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+namespace displays
+{
+JointInfo::JointInfo(const std::string & name, rviz_common::properties::Property * parent_category)
+{
+  name_ = name;
+  effort_ = 0;
+  max_effort_ = 0;
+
+  // info->category_ = new Property( QString::fromStdString( info->name_ ) , QVariant(), "",
+  // joints_category_);
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]: Just delete this commented-out code.

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+    return frame;
+  }
+
+  std::string composite = prefix;
+  composite.append("/");
+  composite.append(frame);
+  return composite;
+}
+
+void EffortDisplay::processMessage(sensor_msgs::msg::JointState::ConstSharedPtr msg)
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp:279]: Need to `#include <sensor_msgs/msg/joint_state.hpp>` at the top of the file for this.

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+{
+  // Robot model might not be loaded already
+  if (!robot_model_) {
+    std::cerr << "Robot model might not be loaded already" << '\n';
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]:
```suggestion
  // Robot model might not be loaded yet
  if (!robot_model_) {
    std::cerr << "Robot model might not be loaded yet" << '\n';
```

Besides this, an unconditional print here seems problematic.  It seems like 1) the user might miss this on the console, and 2) we might spam the user early on with a lot of these messages.  Could we instead update the Status field somehow?

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+    sub_opts.event_callbacks.message_lost_callback =
+      [&](rclcpp::QOSMessageLostInfo & info)
+      {
+        std::ostringstream sstm;
+        sstm << "Some messages were lost:\n>\tNumber of new lost messages: " <<
+          info.total_count_change << " \n>\tTotal number of messages lost: " <<
+          info.total_count;
+        setStatus(
+          rviz_common::properties::StatusLevel::Warn,
+          "Array Topic",
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]: Do you know what "Array Topic" refers to here?

File: rviz_rendering/src/rviz_rendering/objects/effort_visual.cpp [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+{
+EffortVisual::EffortVisual(Ogre::SceneManager * scene_manager, Ogre::SceneNode * parent_node)
+: scene_manager_(scene_manager), parent_node_(parent_node)
+{
+}
+
+void EffortVisual::getRainbowColor(float value, Ogre::ColourValue & color)
+{
+  value = std::min(value, 1.0f);
+  value = std::max(value, 0.0f);
```
Comment (clalancette) [inline rviz_rendering/src/rviz_rendering/objects/effort_visual.cpp:46]: I think you need `#include <algorithm>` at the top of the file for this.

File: rviz_rendering/src/rviz_rendering/objects/effort_visual.cpp [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+
+  float h = value * 5.0f + 1.0f;
+  int i = floor(h);
```
Comment (clalancette) [inline rviz_rendering/src/rviz_rendering/objects/effort_visual.cpp]: I think you need `#include <cmath>` at the top of the file for this.

File: rviz_rendering/test/rviz_rendering/objects/effort_visual_test.cpp [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+ */
+
+#include <gmock/gmock.h>
+
+// #include <memory>
+// #include <string>
+// #include <vector>
+//
+// #include <OgreBillboardChain.h>
+// #include <OgreManualObject.h>
```
Comment (clalancette) [inline rviz_rendering/test/rviz_rendering/objects/effort_visual_test.cpp]: Remove this commented-out code.

File: rviz_rendering/test/rviz_rendering/objects/effort_visual_test.cpp [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+#include <OgreRoot.h>
+#include <OgreSceneNode.h>
+//
+#include "../ogre_testing_environment.hpp"
+#include "../scene_graph_introspection.hpp"
+#include "rviz_rendering/objects/effort_visual.hpp"
+// #include "rviz_rendering/objects/billboard_line.hpp"
+// #include "rviz_rendering/objects/arrow.hpp"
+//
```
Comment (clalancette) [inline rviz_rendering/test/rviz_rendering/objects/effort_visual_test.cpp]: Remove this commented-out code.

File: rviz_rendering/test/rviz_rendering/objects/effort_visual_test.cpp [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+class EffortVisualTestFixture : public ::testing::Test
+{
+protected:
+  void SetUp()
+  {
+    testing_environment_ = std::make_shared<rviz_rendering::OgreTestingEnvironment>();
+    testing_environment_->setUpOgreTestEnvironment();
+  }
+
+  std::shared_ptr<rviz_rendering::OgreTestingEnvironment> testing_environment_;
```
Comment (clalancette) [inline rviz_rendering/test/rviz_rendering/objects/effort_visual_test.cpp:73]: Add `#include <memory>` at the top of the file.

File: rviz_rendering/test/rviz_rendering/objects/effort_visual_test.cpp [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+Ogre::SceneNode * findForceArrow(Ogre::SceneNode * scene_node)
+{
+  auto arrows = rviz_rendering::findAllArrows(scene_node);
+  auto billboard_line = rviz_rendering::findOneBillboardChain(scene_node);
+  for (const auto & arrow : arrows) {
+    if (billboard_line->getParentSceneNode()->getParent() == arrow->getParent()) {
+      return arrow;
+    }
+  }
+  std::cerr << "here" << '\n';
```
Comment (clalancette) [inline rviz_rendering/test/rviz_rendering/objects/effort_visual_test.cpp]: Add `#include <iostream>` at the top of the file.

**Review by clalancette — CHANGES_REQUESTED** [review]
First pass review.

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 20]
```diff
@@ -0,0 +1,419 @@
...
+  setStatus(rviz_common::properties::StatusProperty::Ok, "URDF", "Robot model parsed Ok");
+
+  for (std::map<std::string, urdf::JointSharedPtr>::iterator it = robot_model_->joints_.begin();
+    it != robot_model_->joints_.end(); it++)
+  {
+    urdf::JointSharedPtr joint = it->second;
+    if (joint->type == urdf::Joint::REVOLUTE || joint->type == 2) {
+      std::string joint_name = it->first;
+      urdf::JointLimitsSharedPtr limit = joint->limits;
+      joints_[joint_name] = createJoint(joint_name);
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]: At this point, `createJoint` is a one-liner, so I'll suggest we just inline it here:

```suggestion
      joints_[joint_name] = new JointInfo(joint_name, joints_category_);
```

and remove the the `createJoint` method.

File: rviz_default_plugins/include/rviz_default_plugins/displays/effort/effort_display.hpp [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+  void topic_callback(const std_msgs::msg::String & msg);
+
+  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
+  rclcpp::executors::SingleThreadedExecutor::SharedPtr executor_;
+
+  // Storage for the list of visuals.  It is a circular buffer where
+  // data gets popped from the front (oldest) and pushed to the back (newest)
+  std::deque<std::shared_ptr<rviz_rendering::EffortVisual>> visuals_;
+
+  typedef std::map<std::string, JointInfo *> M_JointInfo;
```
Comment (clalancette) [inline rviz_default_plugins/include/rviz_default_plugins/displays/effort/effort_display.hpp]: I just had a thought about this.  If we instead made this:

```suggestion
  typedef std::map<std::string, std::unique_ptr<JointInfo>> M_JointInfo;
```

Then we wouldn't need to explicitly free it in the destructor.  We'd also have to change where we allocate it to `std::make_unique`, but I think that is probably worth it.

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 22]
```diff
@@ -0,0 +1,419 @@
...
+{
+  rviz_common::Display::load(config);
+}
+
+void EffortDisplay::topic_callback(const std_msgs::msg::String & msg)
+{
+  robot_description_ = msg.data;
+  robot_model_ = std::shared_ptr<urdf::Model>(new urdf::Model());
+  if (!robot_model_->initString(robot_description_)) {
+    // ROS_ERROR("Unable to parse URDF description!");
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]: Remove this commented-out code.

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 23]
```diff
@@ -0,0 +1,419 @@
...
+
+  std::string composite = prefix;
+  composite.append("/");
+  composite.append(frame);
+  return composite;
+}
+
+void EffortDisplay::processMessage(sensor_msgs::msg::JointState::ConstSharedPtr msg)
+{
+  // Robot model might not be loaded already
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]:
```suggestion
  // Robot model might not be loaded yet
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 24]
```diff
@@ -0,0 +1,419 @@
...
+  if (!robot_model_) {
+    std::cerr << "Robot model might not be loaded yet" << '\n';
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]: My previous comment about this still seems to be open:

"an unconditional print here seems problematic. It seems like 1) the user might miss this on the console, and 2) we might spam the user early on with a lot of these messages. Could we instead update the Status field somehow?"

(though maybe this isn't possible)

File: rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp [hunk 25]
```diff
@@ -0,0 +1,419 @@
...
+    return;
+  }
+  // We are keeping a circular buffer of visual pointers.  This gets
+  // the next one, or creates and stores it if the buffer is not full
+  std::shared_ptr<rviz_rendering::EffortVisual> visual;
+  if (visuals_.size() == static_cast<size_t>(history_length_property_->getInt())) {
+    visual = visuals_.front();
+  } else {
+    visual.reset(new rviz_rendering::EffortVisual(context_->getSceneManager(), scene_node_));
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/effort/effort_display.cpp]: I think `std::make_shared` is generally preferred for memory layout and other reasons:

```suggestion
    visual = std::make_shared<rviz_rendering::EffortVisual>(context_->getSceneManager(), scene_node_);
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I left a few more comments; these all should be pretty easy to fix.  Once that is done I'm happy with this.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18993/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13517/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19736/)

**Comment by ahcorde** [comment]
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19743/)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18997/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13521/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19748/)

**Comment by clalancette** [comment]
So the code here is looking good to me.

The thing I'm wondering about is whether we should actually add this to the default set of RViz plugins.  That is, we already have a large set, and it already takes quite a long time to compile this package.  Is this plugin important/universal enough that we should include it in the default set, or can we distribute it in a separate package instead?

**Comment by ahcorde** [comment]
> So the code here is looking good to me.
> 
> The thing I'm wondering about is whether we should actually add this to the default set of RViz plugins. That is, we already have a large set, and it already takes quite a long time to compile this package. Is this plugin important/universal enough that we should include it in the default set, or can we distribute it in a separate package instead?

@clalancette I can add here two arguments:

 - **Out of the box experience**: When people are trying to migrate from ROS 1 to ROS 2, they will expect to have the same behaviour without any additional step.
 - **Feature parity**: With these 3 open PRs we will reach feature parity (finally).

**Comment by ahcorde** [comment]
friendly ping @clalancette

**Comment by clalancette** [comment]
So here is the issue that I'm struggling with this.

First, we've gone many years without these plugins, and haven't had a lot of call for them.  That doesn't mean there has been zero call for them, or that they don't have value, but these PRs are the first time I've heard of them requested.  That makes me think that they aren't necessarily widely used, and hence could be in a separate package.  That would also make it so that these plugins aren't release-critical, and thus keep from adding additional burden on the core maintainers.

Second, our build and test times are already absurd.  Here are some numbers:

* On Linux, it currently takes us 2 hours and 26 minutes to run a [nightly](https://ci.ros2.org/view/nightly/job/nightly_linux_release/2702/) full build and test.  Of that, 59 minutes are spent compiling.  And of that, the largest compile time is for `rviz_default_plugins` at 11 minutes.
* On Windows, the situation is far more dire.  It currently takes us 5 hours and 44 minutes to run a [nightly](https://ci.ros2.org/view/nightly/job/nightly_win_rel/2722/) full build and test.  Of that, 2 hours 45 minutes are spent compiling.  And of that, the largest compile time is for `rviz_default_plugins` at 32 minutes.

This combination of factors makes me think that these plugins should be released into a separate package for manipulation.  Maybe `rviz_manipulation_plugins` or something like that.  It could then be released into humble, iron, and rolling, and be available as normal.

All of that said, I could be convinced otherwise.  I'd also like to hear from @ros2/team here for other opinions.

**Comment by sloretz** [comment]
> [...] And of that, the largest compile time is for rviz_default_plugins at 32 minutes.

I agree `rviz_default_plugins` is really slow to build. By how much does adding this plugin increase the compile time? If it increases it significantly then I'd see that as justification to put it in a different package.

**Comment by ahcorde** [comment]
I created this branch `ahcorde/rolling/rviz_default_plugins_parity` which includes the remaining rviz default plugins to get the feature parity. Let's see how much increase the compilation time.

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19796/)

**Comment by ahcorde** [comment]
@clalancette in this last build the time to compile rviz_default_plugins with these three new plugins is 

```bash
Finished <<< rviz_default_plugins [22min 55s]
```

What do you think?  can we move forward ?

**Comment by clalancette** [comment]
> What do you think? can we move forward ?

Looking at the build time in your example branch, it is actually substantially less than the current nightly CI times ([22 minutes for your CI](https://ci.ros2.org/job/ci_windows/19796/consoleFull) vs [32 minutes for nightly CI](https://ci.ros2.org/view/nightly/job/nightly_win_rel/2731/consoleFull), and they were running on the same AWS instance).  That seems counter-intuitive; we are adding code, so it should be at least the same amount of time, if not longer.  Do you have any idea why that might be?

**Comment by ahcorde** [comment]
> > What do you think? can we move forward ?
> 
> Looking at the build time in your example branch, it is actually substantially less than the current nightly CI times ([22 minutes for your CI](https://ci.ros2.org/job/ci_windows/19796/consoleFull) vs [32 minutes for nightly CI](https://ci.ros2.org/view/nightly/job/nightly_win_rel/2731/consoleFull), and they were running on the same AWS instance). That seems counter-intuitive; we are adding code, so it should be at least the same amount of time, if not longer. Do you have any idea why that might be?

I don´t have a answer for this, but maybe @j-rivero or @nuclearsandwich have one.

**Comment by ahcorde** [comment]
I was discussing this with @j-rivero offline, and these are his thoughts:

 - we need to check if both build are running with the same paremeters
 - changes in the `CMakeLists.txt` seems trivial, he can not explain any reduction in time.
 - There are failing test in CI but I think that are all related to linting. If there is a test failure in something time-consuming that could explain why the CI is running in less time but does not seem to be the case.

Last comment is not relevant because we are just comparing compilation times. I will check that the parameters are the same.

**Comment by ahcorde** [comment]
The only difference is the `CI_CMAKE_BUILD_TYPE` which is setted to `Release`. Launching again the job

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19839/)

**Comment by nuclearsandwich** [comment]
> I don´t have a answer for this, but maybe @j-rivero or @nuclearsandwich have one.

One thing that jumps out at me is the scoped build argument in the CI job

```
 build_args: --event-handlers console_cohesion+ console_package_list+ --cmake-args -DINSTALL_EXAMPLES=OFF -DSECURITY=ON -DAPPEND_PROJECT_NAME_TO_INCLUDEDIR=ON --packages-above-and-dependencies rviz_default_plugins rviz_rendering rviz_common, 
```

With fewer packages being built there's probably more capacity for the rviz_default_plugins build. To get a true apples-to-apples comparison you would need to build the same set of packages as we do in the nightlies.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19126/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13678/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19857/)

File: rviz_default_plugins/CMakeLists.txt [hunk 26]
```diff
@@ -750,6 +752,23 @@ if(BUILD_TESTING)
...
+    TIMEOUT 180)
+  if(TARGET effort_display_visual_test)
+    target_include_directories(effort_display_visual_test PUBLIC
+      test/page_objects
+      ${TEST_INCLUDE_DIRS}
+      ${rviz_visual_testing_framework_INCLUDE_DIRS})
+    target_link_libraries(effort_display_visual_test
+      ${TEST_LINK_LIBRARIES}
+      rviz_visual_testing_framework::rviz_visual_testing_framework)
+  endif()
```
Comment (clalancette) [inline rviz_default_plugins/CMakeLists.txt]: With the latest changes to the CMakeLists.txt, some of this isn't needed anymore:

```suggestion
  ament_add_gtest(effort_display_visual_test
    test/rviz_default_plugins/displays/effort/effort_display_visual_test.cpp
    test/rviz_default_plugins/publishers/effort_publisher.hpp
    test/rviz_default_plugins/page_objects/effort_display_page_object.cpp
    ${SKIP_VISUAL_TESTS}
    TIMEOUT 180)
  if(TARGET effort_display_visual_test)
    target_include_directories(effort_display_visual_test PRIVATE test)
    target_link_libraries(effort_display_visual_test
      rviz_visual_testing_framework::rviz_visual_testing_framework
      rclcpp::rclcpp
      ${sensor_msgs_TARGETS}
    )
  endif()
```

File: rviz_default_plugins/test/rviz_default_plugins/publishers/effort_publisher.hpp [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+#ifndef RVIZ_DEFAULT_PLUGINS__PUBLISHERS__EFFORT_PUBLISHER_HPP_
+#define RVIZ_DEFAULT_PLUGINS__PUBLISHERS__EFFORT_PUBLISHER_HPP_
+
+#include <string>
+#include <chrono>
+
+#include <rclcpp/rclcpp.hpp>
+#include <sensor_msgs/msg/joint_state.hpp>
+
+using namespace std::chrono_literals; //NOLINT
```
Comment (clalancette) [inline rviz_default_plugins/test/rviz_default_plugins/publishers/effort_publisher.hpp]:
```suggestion
using namespace std::chrono_literals;  // NOLINT
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left two small things to fix, then I think I'm happy with this.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19135/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13694/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19870/)

**Review by clalancette — APPROVED** [review]

**Comment by ahcorde** [comment]
https://github.com/Mergifyio backport humble iron

**Comment by zimmerman-pliant** [comment]
Should this work with continuous joints? I've got an application with a continuous rotary joint where effort display could be useful, but the plugin description mentions `revolute` in particular:

https://github.com/ros2/rviz/blob/rolling/rviz_default_plugins/plugins_description.xml#L66

Edit: swapped my joint to `revolute` temporarily in my URDF and got a working visualization, so empirically I guess it only works with `revolute` joints.

**Pull request merged**